ksql-streams/src/main/java/io/confluent/ksql/execution/streams/SinkBuilder.java

```diff
@@ -0,0 +1,132 @@
+
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.Produced;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.To;
+
+public final class SinkBuilder {
+  private SinkBuilder() {
+  }
+
+  public static  <K> void build(
+      final LogicalSchema schema,
+      final Formats formats,
+      final Optional<TimestampColumn> timestampColumn,
+      final String topicName,
+      final KStream<K, GenericRow> stream,
+      final KeySerdeFactory<K> keySerdeFactory,
+      final QueryContext queryContext,
+      final KsqlQueryBuilder queryBuilder
+  ) {
+    final PhysicalSchema physicalSchema = PhysicalSchema.from(schema, formats.getOptions());
+
+    final Serde<K> keySerde = keySerdeFactory.buildKeySerde(
+        formats.getKeyFormat(),
+        physicalSchema,
+        queryContext
+    );
+
+    final Serde<GenericRow> valueSerde = queryBuilder.buildValueSerde(
+        formats.getValueFormat(),
+        physicalSchema,
+        queryContext
+    );
+
+    final int timestampColumnIndex = timestampColumn.map(TimestampColumn::getColumn)
+        .map(c -> schema.findValueColumn(c).orElseThrow(IllegalStateException::new))
+        .map(Column::index)
+        .orElse(-1);
+
+    stream.transform(new TransformTimestamp<>(timestampColumnIndex))
+        .to(topicName, Produced.with(keySerde, valueSerde));
+  }
+
+  static class TransformTimestamp<K>
+      implements TransformerSupplier<K, GenericRow, KeyValue<K, GenericRow>> {
+    private final int timestampColumnIndex;
+
+    TransformTimestamp(final int timestampColumnIndex) {
+      this.timestampColumnIndex = timestampColumnIndex;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+      if (o == null || !(o instanceof TransformTimestamp)) {
+        return false;
+      }
+
+      final TransformTimestamp that = (TransformTimestamp)o;
+      return timestampColumnIndex == that.timestampColumnIndex;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(timestampColumnIndex);
+    }
+
+    @Override
+    public Transformer<K, GenericRow, KeyValue<K, GenericRow>> get() {
+      return new Transformer<K, GenericRow, KeyValue<K, GenericRow>>() {
+        private ProcessorContext processorContext;
+
+        @Override
+        public void init(final ProcessorContext processorContext) {
+          this.processorContext = requireNonNull(processorContext, "processorContext");
+        }
+
+        @Override
+        public KeyValue<K, GenericRow> transform(final K key, final GenericRow row) {
+          if (timestampColumnIndex >= 0 && row.get(timestampColumnIndex) instanceof Long) {
```

Comment:
The `timestampColumnIndex >= 0` bit can be removed if we're not introducing this step if not timestamp column is supplied.

The `row.get(timestampColumnIndex) instanceof Long` bit should be removed as prior code should already have checked the timestamp column exists and is of the right type.

BTW, I believe CTAS/CSAS support `WITH(TIMESTAMP_FORMAT)` (the docs say they do), so the value may well be a string!

Reply:
Done